cats/cats-core/src/main/java/com/netflix/spinnaker/cats/cache/DefaultCacheData.java

```diff
@@ -34,10 +36,27 @@ public DefaultCacheData(String id, Map<String, Object> attributes, Map<String, C
     }
 
     public DefaultCacheData(String id, int ttlSeconds, Map<String, Object> attributes, Map<String, Collection<String>> relationships) {
+        this(id, ttlSeconds, attributes, relationships, Clock.systemDefaultZone());
+    }
+
+    public DefaultCacheData(String id, int ttlSeconds, Map<String, Object> attributes, Map<String, Collection<String>> relationships, Clock clock) {
+        // ensure attributes is non-null and mutable given that `cacheExpiry` will be added
+        attributes = attributes == null ? new HashMap<>() : new HashMap<>(attributes);
+
         this.id = id;
-        this.ttlSeconds = ttlSeconds;
         this.attributes = attributes;
         this.relationships = relationships;
+
+        if (ttlSeconds > 0) {
```

Comment:
Rather than modifying `RedisCache` such that it could fetch `ttl` for each key (performance? complication?) ... I opt'd to store it as an additional attribute and make `DefaultCacheData` smart enough to calculate the `ttl` on-demand.